usecases/webhooks_usecase.go

```diff
@@ -0,0 +1,224 @@
+					SendAttemptCount: 0,
+					DeliveryStatus:   models.Scheduled,
+					OrganizationId:   input.OrganizationId,
+					PartnerId:        input.PartnerId,
+					EventType:        input.EventType,
+					EventData:        input.EventData,
+				})
+			if err != nil {
+				return models.Webhook{}, fmt.Errorf("error while creating webhook: %w", err)
+			}
+			return usecase.webhookRepository.GetWebhook(ctx, tx, webhookId)
+		})
+
+	go func() {
+		ctx := context.WithoutCancel(ctx)
+		ctx, _ = context.WithTimeout(ctx, 10*time.Second)
+		// This function is meant to be executed asynchronously and return way after the request was finished
+		// so we don't return any error
+		usecase.sendWebhook(ctx, webhook, utils.LoggerFromContext(ctx))
+	}()
+
+	return nil
+}
+
+func (usecase WebhooksUsecase) SendWebhooks(
+	ctx context.Context,
+) error {
+	logger := utils.LoggerFromContext(ctx)
+	exec := usecase.executorFactory.NewExecutor()
+
+	pendingWebhooks, err := usecase.webhookRepository.ListWebhooks(ctx, exec, models.WebhookFilters{
+		DeliveryStatus: []models.WebhookDeliveryStatus{models.Scheduled, models.Retry},
+	})
+	if err != nil {
+		return fmt.Errorf("error while listing pending webhooks: %w", err)
+	}
+	logger.InfoContext(ctx, fmt.Sprintf("Found %d webhooks to send", len(pendingWebhooks)))
+	if len(pendingWebhooks) == 0 {
+		return nil
+	}
+
+	var waitGroup sync.WaitGroup
+	// The channel needs to be big enough to store any possible errors to avoid deadlock due to the presence of a waitGroup
+	uploadErrorChan := make(chan error, len(pendingWebhooks))
+	deliveryStatusChan := make(chan models.WebhookDeliveryStatus, len(pendingWebhooks))
+
+	startProcessSendWebhook := func(webhook models.Webhook) {
+		defer waitGroup.Done()
+		logger := logger.With("webhook_id", webhook.Id)
+		deliveryStatus, err := usecase.sendWebhook(ctx, webhook, logger)
+		if err != nil {
+			uploadErrorChan <- err
+		}
+		if deliveryStatus != nil {
+			deliveryStatusChan <- *deliveryStatus
+		}
+	}
+
+	for _, webhook := range pendingWebhooks {
+		waitGroup.Add(1)
+		go startProcessSendWebhook(webhook)
+	}
+
+	waitGroup.Wait()
+	close(uploadErrorChan)
+	close(deliveryStatusChan)
+
+	errorCount := 0
+	var firstError error
+	for err := range uploadErrorChan {
+		errorCount++
+		if firstError == nil {
+			firstError = err
+		}
+	}
+
+	successCount := 0
+	retryCount := 0
+	failedCount := 0
+	for status := range deliveryStatusChan {
+		switch status {
+		case models.Success:
+			successCount++
+		case models.Retry:
+			retryCount++
+		case models.Failed:
+			failedCount++
+		}
+	}
+	logger.InfoContext(ctx, fmt.Sprintf("Webhooks sent: %d success, %d retry, %d failed, %d errors",
+		successCount, retryCount, failedCount, errorCount))
+
+	return firstError
+}
+
+// sendWebhook sends a webhook and updates its status in the database.
+// It returns the delivery status of the webhook and an error if updating the webhook fails.
+func (usecase *WebhooksUsecase) sendWebhook(
```

Comment:
This function should do what it does in a transaction. We can discuss the details if you want.

Reply:
I tried to change for a tx implementation. Let's try this way and if transaction make webhook bulk send slow, we can remove it by leveradging the idempotency key.